### ISSUE TITLE
Function expects shared_ptr, receives derived class and fails

### ISSUE THREAD

**Issue by JoshKlint** [ThePhD/sol2#616]
I am having trouble calling a function that expects a base shared_ptr and receives a derived shared_ptr object. World::Render() expects a shared_ptr(Buffer) and is instead given a shared_ptr(Context), which should still work, since Context is derived from Buffer.

Here are the relevant C++ classes and functions:
```
class World
{
     void Render(shared_ptr<Buffer> buffer);
};
class Buffer {};
class Context : public Buffer {};
extern shared_ptr<Context> CreateContext();
extern shared_ptr<World> CreateWorld();
```
Here is the binding code:
```
lua.new_usertype<World>("World", "Render", &World::Render);
lua.new_usertype<Buffer>("Buffer");
lua.new_usertype<Context>("Context", sol::base_classes, sol::bases<Buffer>);
lua.set_function("CreateContext", CreateContext);
lua.set_function("CreateWorld", CreateWorld);
```
And here is the Lua script:
```
local context = CreateContext()
local world = CreateWorld()
world:Render(context)
```
This results in the following error:
> Lua Error: sol: runtime error: [string "print("Starting...")..."]:17: stack index 2, expected userdata, received sol.sol::detail::unique_usertype<Leadwerks::Context>: unrecognized userdata (not pushed by sol?) (bad argument into 'void(std::shared_ptr<Leadwerks::Buffer>)')
> stack traceback:
>         [C]: in function 'Render'
>         [string "print("Starting...")..."]:17: in main chunk

**Comment by Rochet2** [comment]
I dont think sol supports such (yet).
However if you used `void Render(Buffer& buffer);` for example, it would work.

**Comment by JoshKlint** [comment]
That won't work. We have lots of things like this:

C++:
```
class Entity {};
class Model : public Entity {};
class Sprite: public Entity {};
```
Lua:
```
local model = CreateBox()
local sprite = CreateSprite()
sprite:SetParent(model)-- expects a shared_ptr(Entity), not a shared_ptr(Model)
```

**Comment by ThePhD** [comment]
It is next-to impossible for me to see through `shared_ptr<Derived>` and tell accurately that it can be built from `shared_ptr<Base>`. There has been a long discussion on exactly why this is hard in the first place: 

https://github.com/ThePhD/sol2/issues/505#issuecomment-327033671

I can try to take a crack at this again. I certainly have no expectations I'll come up with anything performant or sane anytime soon, though.

**Comment by ThePhD** [comment]
It's unfortunately a lot more complicated than that. I can know something is `std::shared_ptr<Derived>`. The problem, as I spoke of in the linked issue and the comment trail, is that I have a much harder time identifying that something is just a plain `shared_ptr` or `unique_ptr`, and how to do the conversion from Derived to Base at-runtime. It is going to require either some hackery or finesse.

Note that base classes do work, it's not a problem of figuring out if a class is an Object or not. The problem is knowing that something is a shared_ptr, ripping out the type _inside_ the shared ptr, and performing a proper base-derived shared_ptr conversion.

**Comment by JoshKlint** [comment]
An object stored in a shared_ptr  is going to be created and destroyed infrequently. As a game runs, creating a new entity, material, or other object with some persistent payload is an infrequent event. Adding extra overhead when one of these objects is first exposed to Lua or when it is collected will have no significant impact on performance.

A copy-able object not stored in a shared_ptr, like a Vec3 or Mat4 is something that will be created and destroyed all the time as the game runs. Adding any overhead to these objects would have a significant impact on performance.

Therefore, it would be safe to enter pointers into a map to indicate they are a shared pointer:
```
if (isSharedPointer)
{
    shared_pointers_map[*void] = true;
}
```
And then you could check this table later to see if the object is a shared pointer. Remove the map value when the object is collected from Lua.

**Comment by ThePhD** [comment]
I do not believe this approach is scalable, nor does it help me with the intrinsic problem of properly converting a chunk of memory without violating aliasing rules. The map idea may make sense in your architecture but it likely does not make sense in the general case.

**Comment by JoshKlint** [comment]
Based on an earlier suggestion I tried something new.

Our class hierarchy is like below. Simple classes are passed as objects and all derived from Object, while complex objects are always passed around as shared pointers and are all derived from a SharedObject class (which is also derived from Object):
- Object
  - Vec3
  - Mat4
  - SharedObject
    -World
    -Context
    -Entity
      -Model

The SharedObject class has a special function to get a shared object for itself. This is possible the class has multiple inheritance and is also derived from enable_shared_from_this<T>:
```
	shared_ptr<SharedObject> SharedObject::Self()
	{
		return shared_from_this();
	}
```

Here is our original World::Render() function:
`void World::Render(shared_ptr<Buffer> buffer)`
And now I added a new overload:
```
	void World::Render(Buffer& buffer)
	{
		Render(dynamic_pointer_cast<Buffer>(buffer.Self()));// calls the original function
	}
```
Amazingly, it seems to work. I think this should be safe to use, right?

**Comment by ThePhD** [comment]
This'll work while I figure other things out.

**Comment by ThePhD** [comment]
I'm going to leave this issue open to remind me to do this. I might also ramble about things that do and don't work here.

**Comment by ThePhD** [comment]
The code to fix this is in sol3, but isn't active just yet. Thanks for bringing attention to this so long ago.

**Comment by ThePhD** [comment]
This is now officially fixed and part of sol3, woo! #654 🎉

**Comment by JoshKlint** [comment]
Thanks! I love you!
